### PR TITLE
Per bucket counters

### PR DESCRIPTION
**Pull request by Siddhant085**

### CODE REVIEW

File: core/utils/sys_measure.h [hunk 1]
```diff
@@ -61,6 +66,17 @@ class CoreStats {
...
+  public:
+   BucketStats() {}
+   uint64_t per_bucket_packet_counter[RETA_SIZE] = {0};
+   std::shared_mutex bucket_table_lock;
+   uint32_t rss_hash_to_id(uint32_t hash) {
+     return hash & (RETA_SIZE-1);
+   }
+};
+
+extern BucketStats bucket_stats;
```
Comment (jianfenw) [inline core/utils/sys_measure.h:77]: Add a sentence saying what it does.

File: core/utils/sys_measure.h [hunk 2]
```diff
@@ -3,13 +3,18 @@
...
 
 #include "flow.h"
 #include "lock_less_queue.h"
 
+#define RETA_SIZE 512
 namespace bess{
 namespace utils {
 
+// Since different cores are expected to work on different index we don't
+// synchronize between cores.
```
Comment (jianfenw) [inline core/utils/sys_measure.h]: Why is this comment seperated from BucketStats?

File: core/utils/sys_measure.h [hunk 3]
```diff
@@ -3,13 +3,18 @@
 
 #include <cstdint>
 #include <vector>
+#include <shared_mutex>
```
Comment (jianfenw) [inline core/utils/sys_measure.h:6]: Not required any more?

File: core/modules/nfv_monitor.cc [hunk 4]
```diff
@@ -134,6 +134,17 @@ void NFVMonitor::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
     //   it->second += 1;
     // }
 
+    uint32_t id = bess::utils::bucket_stats.rss_hash_to_id(reinterpret_cast<rte_mbuf*>(pkt)->hash.rss);
+    packet_counter_lock_.lock();
+    bess::utils::bucket_stats.bucket_table_lock.lock_shared();
```
Comment (jianfenw) [inline core/modules/nfv_monitor.cc]: Your comment says: "Access within the same core is synchronized by packet_counter_lock."
Why do you still call lock here? Any cases where multiple NFVMonitor modules run on the same core?

File: core/modules/nfv_monitor.h [hunk 5]
```diff
@@ -77,7 +77,7 @@ class NFVMonitor final : public Module {
   int active_flow_count_;
   float packet_rate_;
   int idle_period_count_;
-
+  std::mutex packet_counter_lock_;
```
Comment (jianfenw) [inline core/modules/nfv_monitor.h]: Not sure why you need this lock. Could you please explain?

File: core/modules/nfv_ctrl.h [hunk 6]
```diff
@@ -31,13 +31,18 @@ class NFVCtrl final : public Module {
...
+  void UpdateFlowAssignment();
 
   CommandResponse CommandGetSummary(const bess::pb::EmptyArg &arg);
 
  private:
   // All available per-core packet queues in a cluster
+  void write_to_gurobi(uint32_t num_cores, std::vector<float> flow_rates, float latency_bound);
   std::vector<WorkerCore> cpu_cores_;
+  uint64_t update_period_;
+  uint64_t last_update_time_;
```
Comment (jianfenw) [inline core/modules/nfv_ctrl.h]: Maybe add a prefix like: long_epoch_update_period_ to your data structs / functions to distinguish from my parts.

File: core/modules/nfv_ctrl.cc [hunk 7]
```diff
@@ -150,6 +155,31 @@ CommandResponse NFVCtrl::Init(const bess::pb::NFVCtrlArg &arg) {
...
+  std::vector<float> flow_rate_per_bucket;
+  int i = 0;
+  bess::utils::bucket_stats.bucket_table_lock.lock();
+  for (i=0; i< RETA_SIZE; i++) {
+    flow_rate_per_bucket.push_back(bess::utils::bucket_stats.per_bucket_packet_counter[i]*1000000000/update_period_);
+    bess::utils::bucket_stats.per_bucket_packet_counter[i] = 0;
+  }
+  bess::utils::bucket_stats.bucket_table_lock.unlock();
+  
+  write_to_gurobi(total_core_count_, flow_rate_per_bucket,slo_p50_);
```
Comment (jianfenw) [inline core/modules/nfv_ctrl.cc]: Okay, this can be complicated. Need to test to make sure that it works.

File: core/modules/nfv_ctrl.cc [hunk 8]
```diff
@@ -132,13 +134,16 @@ CommandResponse NFVCtrl::Init(const bess::pb::NFVCtrlArg &arg) {
   }
 
   total_core_count_ = 0;
+  update_period_ = UPDATE_PERIOD;
+  last_update_time_ = tsc_to_ns(rdtsc());
```
Comment (jianfenw) [inline core/modules/nfv_ctrl.cc]: Add a prefix like:
long_epoch_update_period_
long_epoch_last_update_time_

**Review by jianfenw — COMMENTED** [review]
Thanks, Siddhant. Looks good.
I had several questions about the lock. I wonder if you think it is better to maintain the stats at NFVCore. Then, you only need to protect data from NFVCtrl's access.

File: core/utils/sys_measure.h (hunk 3 above)
Comment (Siddhant085) [reply core/utils/sys_measure.h:6]: Its is used in BucketStats

File: core/modules/nfv_ctrl.cc [hunk 9]
```diff
@@ -150,6 +155,31 @@ CommandResponse NFVCtrl::Init(const bess::pb::NFVCtrlArg &arg) {
 
   return CommandSuccess();
 }
+void NFVCtrl::write_to_gurobi(uint32_t num_cores, std::vector<float> flow_rates, float latency_bound) {
```
Comment (jianfenw) [inline core/modules/nfv_ctrl.cc]: s/write_to_gurobi/WriteToGurobi/

File: core/modules/nfv_ctrl.h [hunk 10]
```diff
@@ -31,13 +31,18 @@ class NFVCtrl final : public Module {
 
   struct task_result RunTask(Context *ctx, bess::PacketBatch *batch, void *arg) override;
   void ProcessBatch(Context *ctx, bess::PacketBatch *batch) override;
+  void UpdateFlowAssignment();
 
   CommandResponse CommandGetSummary(const bess::pb::EmptyArg &arg);
 
  private:
   // All available per-core packet queues in a cluster
+  void write_to_gurobi(uint32_t num_cores, std::vector<float> flow_rates, float latency_bound);
```
Comment (jianfenw) [inline core/modules/nfv_ctrl.h]: Does it have to be a member function?

File: core/modules/nfv_core.cc [hunk 11]
```diff
@@ -289,6 +299,16 @@ void NFVCore::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
       DropPacket(ctx, pkt);
       continue;
     }
+    // Update RSS bucket packet counters.
+    uint32_t id = bess::utils::bucket_stats.rss_hash_to_id(reinterpret_cast<rte_mbuf*>(pkt)->hash.rss);
+    bess::utils::bucket_stats.bucket_table_lock.lock_shared();
+    /*
+     * Access within the same core is synchronized by packet_counter_lock.
```
Comment (jianfenw) [inline core/modules/nfv_core.cc:296]: Delete this.

File: core/modules/nfv_core.cc [hunk 12]
```diff
@@ -229,6 +229,16 @@ void NFVCore::UpdateEpochStats(bess::PacketBatch *batch) {
     if (it == epoch_flow_cache_.end()) {
       epoch_flow_cache_.emplace(flow, true);
     }
+    // Update RSS bucket packet counters.
```
Comment (jianfenw) [inline core/modules/nfv_core.cc]: No need to call it twice.

**Review by jianfenw — COMMENTED** [review]
Looks good.

**Comment by Siddhant085** [comment]
Made the requested changes. I couldn't test the changes as I don't have the updates in nfv_ctrl.cc which call ProcessBatch.

File: core/utils/sys_measure.h [hunk 13]
```diff
@@ -61,6 +63,18 @@ class CoreStats {
   std::vector<Flow> bursty_flows;
 };
 
+class BucketStats {
+  public:
+   BucketStats() {}
+   uint64_t per_bucket_packet_counter[RETA_SIZE] = {0};
+   std::shared_mutex bucket_table_lock;
+   uint32_t rss_hash_to_id(uint32_t hash) {
```
Comment (jianfenw) [inline core/utils/sys_measure.h]: Change this function name before your merge the code.

**Review by jianfenw — APPROVED** [review]
Please change the function name. I will then merge the code for testing.

**Pull request merged**